src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -512,4 +513,64 @@ object messages {
         |}""".stripMargin
   }
 
+  case class WrongNumberOfArgs(fntpe: Type, argKind: String, expectedArgs: List[TypeParamInfo], actual: List[untpd.Tree])(implicit ctx: Context)
+    extends Message(18) {
+    val kind = "Syntax"
+    val expectedCount = expectedArgs.length
+    val actualCount = actual.length
+    val msgPrefix = if (actualCount > expectedCount) "Too many" else "Not enough"
+
+    val expectedArgString = (fntpe.widen match {
+      case pt: MethodOrPoly => pt //ensure we return a type that will have useful typeParms
+      case _ => fntpe
+    }).typeParams.map(_.paramName.show.split("\\$").last).mkString("[", ", ", "]")
```

Comment:
As much as possible, we should avoid playing with regexps on names in error messages, this way lies madness. If there are issues with the way we print names, they should be fixed in a common place, not worked around in every error message.
A slightly nicer way:
```scala
import NameOps._
//...
}).typeParams.map(_.paramName.unexpandedName.show).....
```
Technically this is still incorrect, `unexpandedName` should only be called on a name of a symbol with the flag `ExpandedName` but we have no way of checking that here, I think the correct solution would be to add a `def simpleParamName` method to `TypeParamInfo`, but this could be mentioned in a `TODO` comment and done later.